src/storage/ynab.ts

```diff
@@ -0,0 +1,103 @@
+import { YNAB_TOKEN, YNAB_BUDGET_ID, YNAB_ACCOUNTS } from "../config.js";
+import { SaveStats, TransactionRow, TransactionStorage } from "../types.js";
+import { createLogger } from "./../utils/logger.js";
+import { parseISO, format } from "date-fns";
+import * as ynab from "ynab";
+import hash from "hash-it";
+
+const YNAB_DATE_FORMAT = "yyyy-MM-dd";
+let ynabAPI: ynab.API;
+
+const logger = createLogger("YNABStorage");
+
+export class YNABStorage implements TransactionStorage {
+  async init() {
+    logger("init");
+  }
+
+  canSave() {
+    return Boolean(YNAB_TOKEN && YNAB_BUDGET_ID);
+  }
+
+  async saveTransactions(txns: Array<TransactionRow>) {
+    await this.init();
+
+    ynabAPI = new ynab.API(YNAB_TOKEN);
+    const budgetName = await getBudgetName(YNAB_BUDGET_ID);
+
+    // Convert transactions to YNAB format
+    logger("transforming transactions to ynab format");
+    const transactionsFromFinancialAccount = txns.map(
+      convertTransactionToYnabFormat,
+    );
+
+    // Send transactions to YNAB
+    logger(`sending to YNAB budget: "${budgetName}"`);
+    await ynabAPI.transactions.createTransactions(YNAB_BUDGET_ID, {
+      transactions: transactionsFromFinancialAccount,
+    });
+    logger("transactions sent to YNAB successfully!");
+
+    const stats: SaveStats = {
+      name: "YNABStorage",
+      table: `budget: "${budgetName}"`,
+      total: txns.length,
+      added: txns.length,
+      pending: NaN,
+      skipped: 0,
+      existing: NaN,
+    };
+
+    return stats;
+  }
+}
+
+function convertTransactionToYnabFormat(
+  tx: TransactionRow,
+): ynab.SaveTransaction {
+  const amount = Math.round(tx.chargedAmount * 1000);
+
+  return {
+    account_id: getYnabAccountIdByAccountNumberFromTransaction(tx.account),
+    date: format(parseISO(tx.date), YNAB_DATE_FORMAT, {}),
+    amount,
+    payee_id: null,
+    payee_name: tx.description,
+    cleared: ynab.TransactionClearedStatus.Cleared,
+    approved: false,
+    import_id: hash(tx.hash).toString(),
```

Comment:
Looking at the [ynab api docs](https://github.com/dmlerner/ynab-api/blob/master/docs/TransactionSummary.md#:~:text=%5Boptional%5D-,import_id,-str%2C%20none_type), this filed can be any string and since `tx.hash` is already unique by account and we don't need to hash it, allowing us to remove the dependancy on the `hash-it` package. 

If we add the hash, let's also prefix it with `moneyman` in case its later visible to the user.

```suggestion
    import_id: `moneyman:${tx.hash}`,
```

Reply:
`moneyman` creates a unique id in the `transactionHash` function but this unique string can be longer than 64 characters that is the maximum can be exported to YNAB in the import_id field (this is not documented, by export fails with import_ids longer than 64 characters).

Hashing the hash (I know...) creates a unique string shorter than 64. So... adding the moneyman string is not relevant. In any case, nothing is lost. All the data in the hash (unique id) is also the data of the transaction.

Reply:
In addition to that, the uniqueness of the import_id is the only thing that prevent the same transactions will be imported again and again. YNAB don't care about duplications putting the responsibility in the exporter.

Reply:
@daniel-hauser I think is better continue with the "double hash" that is really a hash of an unique-id.

Reply:
Since ynab has a limit i see no other option. the main motivation was to skip the addition of the `hash-it` dependency, after digging in their code i think its ok